src/middlewared/middlewared/plugins/vm.py

```diff
@@ -929,23 +929,26 @@ def random_mac(self):
         else:
             return update_devices
 
-    @accepts(Int('id'), Bool('disks', default=False))
-    async def do_delete(self, id, disks):
+    @accepts(Int('id'), Bool('zvols', default=False))
+    async def do_delete(self, id, zvols):
         """Delete a VM."""
         status = await self.status(id)
         if isinstance(status, dict):
             if status.get('state') == 'RUNNING':
                 await self.stop(id)
-        if disks:
+        if zvols:
             devices = await self.middleware.call('vm.device.query', [
                 ('vm', '=', id), ('dtype', '=', 'DISK')
             ])
 
-            for device in devices:
+            for zvol in devices:
+                if not device['attributes']['path'].startswith('/dev/zvol/'):
+                    continue
+
                 disk_name = device['attributes']['path'].rsplit(
                     '/dev/zvol/'
                 )[-1]
-                await self.middleware.call('zfs.dataset.delete', disk_name)
+                await self.middleware.call('pool.dataset.delete', disk_name)
```

Comment:
It won't work, this method checks for dataset dependencies and it has at least one (VM). You need to remove from vm.device first, then call this method, if it raises CallError, then it means other attachments exist (e.g. iSCSI) and what to do in this case is undefined.

Reply:
How about we just use zfs.dataset.delete for now?
If the user is using the same zvol for another VM or iSCSI (both highly unadvised), we ignored it.

May revisit this if it becomes an issue in the future?